lib/dart/lib/src/browser/t_web_socket.dart

```diff
@@ -0,0 +1,138 @@
+
+library thrift.src.browser;
+
+import 'dart:async';
+import 'dart:html' show CloseEvent;
+import 'dart:html' show Event;
+import 'dart:html' show MessageEvent;
+import 'dart:html' show WebSocket;
+
+import 'package:crypto/crypto.dart' show CryptoUtils;
+import 'package:thrift/thrift.dart';
+
+/// A [TSocket] backed by a [WebSocket] from dart:html
+class TWebSocket implements TSocket {
+  final Uri url;
+
+  final StreamController<TSocketState> _onStateController;
+  Stream<TSocketState> get onState => _onStateController.stream;
+
+  final StreamController<Object> _onErrorController;
+  Stream<Object> get onError => _onErrorController.stream;
+
+  final StreamController<List<int>> _onMessageController;
+  Stream<List<int>> get onMessage => _onMessageController.stream;
+
+  final List<Completer<List<int>>> _completers = [];
+  final List<_Request> _requests = [];
+
+  TWebSocket(this.url)
+      : _onStateController = new StreamController.broadcast(),
+        _onErrorController = new StreamController.broadcast(),
+        _onMessageController = new StreamController.broadcast() {
+    if (url == null || !url.hasAuthority || !url.hasPort) {
+      throw new ArgumentError("Invalid url");
+    }
+  }
+
+  WebSocket _socket;
+
+  bool get isOpen => _socket != null && _socket.readyState == WebSocket.OPEN;
+
+  bool get isClosed =>
+      _socket == null || _socket.readyState == WebSocket.CLOSED;
+
+  Future open() async {
+    if (!isClosed) {
+      throw new TTransportError(
+          TTransportErrorType.ALREADY_OPEN, "Socket already connected");
+    }
+
+    _socket = new WebSocket(url.toString());
+    _socket.onError.listen(_onError);
+    _socket.onOpen.listen(_onOpen);
+    _socket.onClose.listen(_onClose);
+    _socket.onMessage.listen(_onMessage);
+  }
+
+  Future close() async {
+    if (_socket != null) {
+      _socket.close();
+    }
+  }
+
+  Future<List<int>> send(List<int> data) {
+    Completer<List<int>> completer = new Completer();
+
+    _requests.add(new _Request(data, completer));
+    _sendRequests();
+
+    return completer.future;
+  }
+
+  void _sendRequests() {
+    while (isOpen && _requests.isNotEmpty) {
+      _Request request = _requests.removeAt(0);
+      _completers.add(request.completer);
+      _socket.sendString(CryptoUtils.bytesToBase64(request.data));
+    }
+  }
+
+  void _onOpen(Event event) {
+    _onStateController.add(TSocketState.OPEN);
+    _sendRequests();
+  }
+
+  void _onClose(CloseEvent event) {
+    _socket = null;
+    _completers.clear();
+    _requests.clear();
+
+    _onStateController.add(TSocketState.CLOSED);
+  }
+
+  void _onMessage(MessageEvent event) {
+    List<int> data;
+
+    try {
+      data = CryptoUtils.base64StringToBytes(event.data);
+    } on FormatException catch (_) {
+      _onErrorController
+          .add(new UnsupportedError("Expected a Base 64 encoded string."));
+    }
+
+    if (!_completers.isEmpty) {
+      _completers.removeAt(0).complete(data);
```

Comment:
Just to make sure I understand this correctly, is this assuming that every request sent to the server will be accompanied by a response? Is it actually guaranteed that they will come back in the same order? Or do they need to be mapped to an ID?